blog/modules/ROOT/pages/6-garbage-collector-perf.adoc

```diff
@@ -0,0 +1,595 @@
+   spend all its time running GC cycles. Garbage collectors therefore _need_
+   excess memory to work with, on top of the memory you would expect to need to allocate
+   all the objects in your program.
+
+Even from this theoretical analysis, we have already found a number of surprising results
+in how GCs perform over time.
+
+## Java Garbage Collectors
+
+Now that we have run through a theoretical introduction and analysis of how GCs work
+and how we would expect them to perform, let's look at some small Java programs and
+monitor how garbage collection happens when using them:
+
+```java
+public class GC {
+    public static void main(String[] args) throws Exception{
+        final int liveSetSize = Integer.parseInt(args[0]);
+        final int maxObjectSize = 200;
+        final int garbageRate = Integer.parseInt(args[1]);
+        Object[] liveSet = new Object[liveSetSize];
+        for(int i = 0; i < liveSetSize; i++) liveSet[i] = new Object[i % maxObjectSize];
+
+        long maxPause = 0;
+        long lastPrintTime = System.currentTimeMillis();
+        int accumulator = 0;
+
+        while(Math.abs(1) == 1){
+            Thread.sleep(1);
+            long loopStartTime = System.currentTimeMillis();
+
+            for(int i = 0; i < garbageRate; i++){
+                int liveSetIndex = (accumulator * garbageRate + i) % liveSetSize;
+                if (liveSetIndex < 0) liveSetIndex *= -1;
+                liveSet[liveSetIndex] = new Object[i % maxObjectSize];
+                accumulator += liveSet[liveSetIndex].hashCode();
+            }
+            long loopTime = System.currentTimeMillis() - loopStartTime;
+
+            if (loopTime > maxPause) maxPause = loopTime;
+
+            long now = System.currentTimeMillis();
+            if (now - lastPrintTime > 5000){
+                lastPrintTime = now;
+                System.out.println("longest GC: " + maxPause + "ms");
+                maxPause = 0;
+            }
+        }
+        System.out.println(accumulator);
+        System.out.println(liveSet);
+    }
+}
+```
+
+This is a small Java program that starts off allocating a bunch of `Object[]` arrays in `liveSet`,
+and then in a loop continuously allocates more ``Object[]``s and over-writes the references
+to older ones. We cycle through the slots in `liveSet` such objects stay around a bit
+before becoming unreachable (garbage). Lastly, we use a few oddities like
+`while (Math.abs(1) == 1)` and `System.out.println` to prevent the JVM from optimizing
+it all away.
+
+This is a bit of a synthetic benchmark, and much less complex than a real-world
+application would be, but it is enough for the purposes of this article.
+
+You can run this program with different Java garbage collectors, heap sizes, `liveSet` size and `garbageRate`
+via:
+
+```bash
+> javac GC.java
+> java -Xmx1g -XX:+UseParallelGC GC 500000 200
+> java -Xmx1g -XX:+UseConcMarkSweepGC GC 500000 200
```

Comment:
Not sure if having CMS still here in the example is that great. Its not shipped in newer JDKs.

So, maybe other options like the low latency collector is better?

Reply:
It would be better to have ZGC generational too.
CMS is only for Java 8.